Fixed docstring of depolarising

RohitP2005 opened the pull request:
## Description
Related to: #1085 
This PR fixes the docstring for the depolarizing channel implementation in toqito incorrectly applies the parameter param_p, swapping p and 1 - p compared to the standard convention. Additionally, the documentation lacks clarity on how the parameter is applied, which needs to be addressed.

## Changes
docstring updated

## Checklist
Before marking your PR ready for review, make sure you checked the following locally. If this is your first PR, you might be notified of some workflow failures after a maintainer has approved the workflow jobs to be run on your PR. 

Additional information is available in the [documentation](https://toqito.readthedocs.io/en/latest/contributing.html#testing).

  -  [x] Use `ruff` for errors related to code style and formatting.
  -  [x] Verify all previous and newly added unit tests pass in `pytest`.
  -  [ ] Check the documentation build does not lead to any failures. `Sphinx` build can be checked locally for any failures related to your PR
  -  [ ] Use `linkcheck` to check for broken links in the documentation
  -  [ ] Use `doctest` to verify the examples in the function docstrings work as expected.

RohitP2005 commented:
Hey @vprusso I think the changes made address the issue.
Is there any other documentation that needs to be updated

purva-thakre commented:
@RohitP2005 No, they don't. You need to fix the test failures too.

RohitP2005 commented:
> @RohitP2005 No, they don't. You need to fix the test failures too.

I will fix the test with the upcomming commit
can u brief on the documentation update that i need to perform

purva-thakre commented:
I don't understand what you are asking here. You can see the failures below

![image](https://github.com/user-attachments/assets/16b2f0f5-0e47-4fb6-9c71-733ec7f81e55)

RohitP2005 commented:
Is there anything else I have to fix other than the tests in this PR

vprusso commented:
@RohitP2005 it migth be good practice as well to flag this documentation clarification on the QETLAB issue board as well.

vprusso commented on toqito/channels/depolarizing.py:
```diff
@@ -84,12 +84,20 @@ def depolarizing(dim: int, param_p: float = 0) -> np.ndarray:
+
+    Note:
+        This function follows the standard convention where `param_p` controls the amount of
+        depolarization applied. If using the Qetlab convention where `param_p` and `1 - param_p`
```
```suggestion
        depolarization applied. If using the QETLAB convention where `param_p` and `1 - param_p`
```

vprusso commented on toqito/channels/depolarizing.py:
```diff
@@ -84,12 +84,20 @@ def depolarizing(dim: int, param_p: float = 0) -> np.ndarray:
 
+
+    Note:
+        This function follows the standard convention where `param_p` controls the amount of
```
We should also add a reference to QETLAB in this doc-string.

RohitP2005 replied:
Okay , will be added to it

vprusso replied:
It seems like you have not added this yet, right?

purva-thakre commented:
@RohitP2005 Are you planning to complete this?

RohitP2005 commented:
Hey @vprusso  could u point out the section of dcostring that is failing . I couldnt find it locally

vprusso commented:
@RohitP2005 You can see the failing tests that are linked in this thread:
<img width="942" alt="Screenshot 2025-04-06 at 11 31 37 AM" src="https://github.com/user-attachments/assets/01a734e9-cc10-4117-b7e4-17e979262fe6" />


https://github.com/vprusso/toqito/actions/runs/14293691919/job/40058156041?pr=1106

RohitP2005 commented:
I'm aware of the failing tests in the thread. What I need help with is identifying the specific section of the docstring code that's causing the issue. If you could point me to the location in the codebase where that problematic docstring lives, that would really help me resolve it quickly.

RohitP2005 commented:
Hey @vprusso anything else that has to be addressed ?

vprusso commented on docs/intro_tutorial.rst:
```diff
@@ -475,7 +475,7 @@ where :math:`\mathbb{I}` is the identity operator and :math:`d` is the dimension
     >>> 
     >>> # Generate the depolarizing channel Choi matrix with noise probability p = 0.3.
-    >>> choi = depolarizing(2, 0.3)
+    >>> choi = depolarizing(2, param_p=0.7)
```
Is there a reason that the argument value was changed?

RohitP2005 replied:
The test seems to fail after the changes made in the depolarizing function

vprusso replied:
Was it failing due to the parameter not being specified or the change in value? And what was the failure?

vprusso commented on toqito/channel_metrics/diamond_distance.py:
```diff
@@ -21,7 +21,7 @@ def diamond_distance(choi_1: np.ndarray, choi_2: np.ndarray) -> float:
     >>> from toqito.channels import depolarizing
     >>> from toqito.channel_metrics import diamond_distance
-    >>> choi_depolarizing = depolarizing(dim=2, param_p=0.2)
+    >>> choi_depolarizing = depolarizing(dim=2, param_p=1)
```
Same here, just wondering why this change was made to the value.

RohitP2005 replied:
It caused the the docstring to fail

vprusso replied:
Can you elaborate a bit on this? What was the failure? Was it that `param_p=0.2` was failing? If so, how, and why does it work for `param_p=1`?

vprusso commented on toqito/channel_ops/tests/test_partial_channel.py:
```diff
@@ -109,13 +109,14 @@
     """Test function works as expected for valid inputs."""
+    channel = depolarizing(2, param_p=1)
+
     if sys_arg is None and dim_arg is None:
```
Is it necessary to have a if else block for this? Wouldn't the function defaults handle any of these cases?

RohitP2005 replied:
case where ```sys_arg``` is none was not handled which made the tests fails

vprusso replied:
Yes, but it has a default argument of `None`. So I don't see why you need the if/else logic in the test. If you don't supply `sys_arg`, by the default, the argument here will be `None`, right? So is it necessary to have branching logic in your test if this is already handled by the function?

vprusso replied:
Seems like this is still unresolved.

vprusso commented on toqito/channel_props/is_unitary.py:
```diff
@@ -54,7 +54,7 @@ def is_unitary(phi: np.ndarray | list[list[np.ndarray]]) -> bool:
     >>> from toqito.channels import depolarizing
     >>> from toqito.channel_props import is_unitary
-    >>> is_unitary(depolarizing(2))
+    >>> is_unitary(depolarizing(2, param_p=1))
```
Is there a reason that the parameter is added here instead of using the default as was done before?

vprusso replied:
> Test was failing due to the missing keyword argument

But this doesn't make sense. There is a default value of `0`. 

https://github.com/vprusso/toqito/blob/1379480fc43a9123e141fbd0794eaa8851cb6ec9/toqito/channels/depolarizing.py#L8

Why would this fail, and if so, what was the failure?

vprusso commented on toqito/channels/depolarizing.py:
```diff
@@ -84,12 +85,25 @@ def depolarizing(dim: int, param_p: float = 0) -> np.ndarray:
+        are swapped, adjust your input accordingly.
+
+        The convention used here is consistent with standard quantum information references
+        such as Nielsen & Chuang. However, it differs from the convention used by QETLAB
```
Please make use of the refs.bib file to explicitly reference these items (i.e. Nielsen and Chuang, QETLAB, etc)

RohitP2005 replied:
Yeah, will do it in the next commit

vprusso commented:
Hey @RohitP2005 

> Can u perform a squash merge on this

Not sure I follow, wouldn't you do this on your branch?

vprusso commented on docs/intro_tutorial.rst:
```diff
@@ -845,4 +845,4 @@ References
 .. bibliography:: 
-    :filter: docname in docnames
\ No newline at end of file
+    :filter: docname in docnames
```
Seems odd this is being flagged as a change. No action, just an observation.

RohitP2005 commented:
```  
  >>> choi = depolarizing(2, 0.3)
  >>> choi = depolarizing(2, param_p=0.7)
```
this caused error due to both missing keyword parameter and change in value
shall i document the use keyword parameter

vprusso commented:
> ```
>   >>> choi = depolarizing(2, 0.3)
>   >>> choi = depolarizing(2, param_p=0.7)
> ```
> 
> this caused error due to both missing keyword parameter and change in value shall i document the use keyword parameter

Could you show me the error though? I'm a bit confused, as the function prototype is:

```py
def depolarizing(dim: int, param_p: float = 0)
```
which has a default value for `param_p`.

RohitP2005 commented:
```
Trying:
    output_state = apply_channel(rho, choi)
Expecting nothing
ok
Trying:
    print(output_state)
Expecting:
    [[0.65 0.  ]
     [0.   0.35]]
**********************************************************************
File "./docs/intro_tutorial.rst", line 482, in intro_tutorial.rst
Failed example:
    print(output_state)
Expected:
    [[0.65 0.  ]
     [0.   0.35]]
Got:
    [[0.85 0.  ]
     [0.   0.15]]
Trying:
    import numpy as np
Expecting nothing
ok
```

vprusso commented:
> ```
> Trying:
>     output_state = apply_channel(rho, choi)
> Expecting nothing
> ok
> Trying:
>     print(output_state)
> Expecting:
>     [[0.65 0.  ]
>      [0.   0.35]]
> **********************************************************************
> File "./docs/intro_tutorial.rst", line 482, in intro_tutorial.rst
> Failed example:
>     print(output_state)
> Expected:
>     [[0.65 0.  ]
>      [0.   0.35]]
> Got:
>     [[0.85 0.  ]
>      [0.   0.15]]
> Trying:
>     import numpy as np
> Expecting nothing
> ok
> ```

But this is a docstring error that isn't for the depolarizing channel. It seems to be a different error for a different function for a different reason.

RohitP2005 commented:
```
.. code-block:: python

    >>> import numpy as np
    >>> from toqito.states import basis
    >>> from toqito.channel_ops import apply_channel
    >>> from toqito.channels import depolarizing
    >>> 
    >>> # # Create a quantum state |0⟩⟨0|.
    >>> rho = np.array([[1, 0], [0, 0]])
    >>> 
    >>> # Generate the depolarizing channel Choi matrix with noise probability p = 0.3.
    >>> choi = depolarizing(2, param_p=0.3)
    >>> 
    >>> # Apply the depolarizing channel using apply_channel.
    >>> output_state = apply_channel(rho, choi)
    >>> print(output_state)
    [[0.65 0.  ]
     [0.   0.35]]
```
this is the section of code that give the docstring failure

RohitP2005 commented:
> Hey @RohitP2005
> 
> > Can u perform a squash merge on this
> 
> Not sure I follow, wouldn't you do this on your branch?

I will take care of it .

vprusso commented:
> this is the section of code that give the docstring failure
> 
> https://github.com/vprusso/toqito/blob/a541df2e7de73f70a83bc60245d9f9b24892ca44/docs/intro_tutorial.rst?plain=1#L478

Yes, but I think the only reason for the failure here is that the output of the matrix is different. This makes sense as the `param_p` parameter is different. So I think one way to fix would simply be to match the `Got:` section with what the output here should be. Does that make sense?

RohitP2005 commented:
> > this is the section of code that give the docstring failure
> > https://github.com/vprusso/toqito/blob/a541df2e7de73f70a83bc60245d9f9b24892ca44/docs/intro_tutorial.rst?plain=1#L478
> 
> Yes, but I think the only reason for the failure here is that the output of the matrix is different. This makes sense as the `param_p` parameter is different. So I think one way to fix would simply be to match the `Got:` section with what the output here should be. Does that make sense?

Got it @vprusso will make the changes in the next commit

vprusso reviewed (approved):
I think this LGTM, but I'll hold off on merging until @purva-thakre can take a look (which likely won't be until the end of this week). Thank you, @RohitP2005 !

purva-thakre commented on toqito/channels/depolarizing.py:
```diff
@@ -69,11 +69,12 @@ def depolarizing(dim: int, param_p: float = 0) -> np.ndarray:
-           [ 2.5 ,  7.25,  3.5 ,  4.  ],
-           [ 4.5 ,  5.  ,  9.75,  6.  ],
-           [ 6.5 ,  7.  ,  7.5 , 12.25]])
+    >>> apply_channel(test_input_mat, depolarizing(4, param_p=1))
```
I don't understand why you changed `param_p = 1` for all the examples?

purva-thakre replied:
never mind, I'll take the rest of the issue from here.

vprusso replied:
But it doesn't require an argument--a default is provided:
https://github.com/vprusso/toqito/blob/1379480fc43a9123e141fbd0794eaa8851cb6ec9/toqito/channels/depolarizing.py#L8

RohitP2005 commented:
@vprusso @purva-thakre  If there's anything that needs to be handled in this issue, i happy to work on it.

vprusso reviewed (commented):
Hi @RohitP2005 

I added some additional comments. Thank you for looking into this and let me know if any of my comments are unclear!

purva-thakre commented:
@vprusso For all the `param_p = 1` issues, I am trying to debug them in #1160 and #1159. We could close this PR. 

I have moved OP's commits to the other PR too. 

The failures in the other PR are because I need to fix some dephasing channel errors.

vprusso commented:
> @vprusso For all the `param_p = 1` issues, I am trying to debug them in #1160 and #1159. We could close this PR.
> 
> I have moved OP's commits to the other PR too.
> 
> The failures in the other PR are because I need to fix some dephasing channel errors.

Ah okay, I see. That makes sense to me then. Does that seem reasonable as well to you, @RohitP2005 ? You are welcome to look at the linked issues as well as I'm sure your insight there would be valuable!

RohitP2005 commented:
> > @vprusso For all the `param_p = 1` issues, I am trying to debug them in #1160 and #1159. We could close this PR.
> > I have moved OP's commits to the other PR too.
> > The failures in the other PR are because I need to fix some dephasing channel errors.
> 
> Ah okay, I see. That makes sense to me then. Does that seem reasonable as well to you, @RohitP2005 ? You are welcome to look at the linked issues as well as I'm sure your insight there would be valuable!


> Hi @RohitP2005
> 
> I added some additional comments. Thank you for looking into this and let me know if any of my comments are unclear!

The doctest fails because param_p defaults to 0, which applies the identity channel (no depolarization), returning the input unchanged. We should either set param_p=1 in the example or update the expected output to match the identity behavior.

Kindly could you please use the "Review changes" option on GitHub when leaving comments? That way I’ll be able to reply directly to your feedback. Thanks!

RohitP2005 commented:
> > @vprusso For all the `param_p = 1` issues, I am trying to debug them in #1160 and #1159. We could close this PR.
> > I have moved OP's commits to the other PR too.
> > The failures in the other PR are because I need to fix some dephasing channel errors.
> 
> Ah okay, I see. That makes sense to me then. Does that seem reasonable as well to you, @RohitP2005 ? You are welcome to look at the linked issues as well as I'm sure your insight there would be valuable!

Yes, I would love to help on that

vprusso commented:
> Yes, I would love to help on that

That would be great! Feel free to comment on the other issues that @purva-thakre linked!

purva-thakre commented:
> The doctest fails because param_p defaults to 0, which applies the identity channel (no depolarization), returning the input unchanged. We should either set param_p=1 in the example or update the expected output to match the identity behavior.

@RohitP2005 Yes, I am aware of this. But when I was looking at some of the outputs of `channel_metrics` functions, I noticed their value was greater than 1. Instead of making you perform some changes, I preferred to make some changes myself. 

I need to double-check the definitions of the metric functions in toqito and add a check to verify that the outputs are bounded within a range. If they are not within the desired range, I want to raise an error. When I create the issue to make these changes, you are more than welcome to work on it, if you want to. 

I am closing this PR in favor of #1160. As I said in an earlier reply, your commits from this PR were already carried over to the other PR.

RohitP2005 commented:
> > The doctest fails because param_p defaults to 0, which applies the identity channel (no depolarization), returning the input unchanged. We should either set param_p=1 in the example or update the expected output to match the identity behavior.
> 
> @RohitP2005 Yes, I am aware of this. But when I was looking at some of the outputs of `channel_metrics` functions, I noticed their value was greater than 1. Instead of making you perform some changes, I preferred to make some changes myself.
> 
> I need to double-check the definitions of the metric functions in toqito and add a check to verify that the outputs are bounded within a range. If they are not within the desired range, I want to raise an error. When I create the issue to make these changes, you are more than welcome to work on it, if you want to.
> 
> I am closing this PR in favor of #1160. As I said in an earlier reply, your commits from this PR were already carried over to the other PR.

@vprusso thanks for heads up!
yes i am looking forward to work on the issue
kindly mention me when u open the issue
